// csv-writer-portable/dist/lib/csv-writer-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CsvWriterFactory = void 0;
var csv_writer_1 = require("./csv-writer");
var CsvWriterFactory = /** @class */ (function () {
    function CsvWriterFactory(csvStringifierFactory) {
        this.csvStringifierFactory = csvStringifierFactory;
    }
    CsvWriterFactory.prototype.createArrayCsvWriter = function (params) {
        var _a;
        var csvStringifier = this.csvStringifierFactory.createArrayCsvStringifier({
            header: params.header,
            fieldDelimiter: params.fieldDelimiter,
            recordDelimiter: params.recordDelimiter,
            alwaysQuote: params.alwaysQuote,
            quoteEmptyFields: params.quoteEmptyFields,
            filterFunction: params.filterFunction,
        });
        return new csv_writer_1.CsvWriter(csvStringifier, params.path, params.encoding, params.append, (_a = params.useBom) !== null && _a !== void 0 ? _a : false);
    };
    CsvWriterFactory.prototype.createObjectCsvWriter = function (params) {
        var _a;
        var csvStringifier = this.csvStringifierFactory.createObjectCsvStringifier({
            header: params.header,
            fieldDelimiter: params.fieldDelimiter,
            recordDelimiter: params.recordDelimiter,
            headerIdDelimiter: params.headerIdDelimiter,
            alwaysQuote: params.alwaysQuote,
            quoteEmptyFields: params.quoteEmptyFields,
            filterFunction: params.filterFunction,
        });
        return new csv_writer_1.CsvWriter(csvStringifier, params.path, params.encoding, params.append, (_a = params.useBom) !== null && _a !== void 0 ? _a : false);
    };
    return CsvWriterFactory;
}());
exports.CsvWriterFactory = CsvWriterFactory;
