// csv-writer-portable/dist/lib/field-stringifier.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldStringifier = exports.VALID_FIELD_DELIMITERS = exports.DEFAULT_FIELD_DELIMITER = void 0;
exports.createFieldStringifier = createFieldStringifier;
exports.DEFAULT_FIELD_DELIMITER = ',';
exports.VALID_FIELD_DELIMITERS = [exports.DEFAULT_FIELD_DELIMITER, ';', '|', '\t'];
var FieldStringifier = /** @class */ (function () {
    function FieldStringifier(fieldDelimiter, quoteEmptyFields, filterFunction) {
        if (quoteEmptyFields === void 0) { quoteEmptyFields = false; }
        if (filterFunction === void 0) { filterFunction = function (value) {
            return value;
        }; }
        this.fieldDelimiter = fieldDelimiter;
        this.quoteEmptyFields = quoteEmptyFields;
        this.filterFunction = filterFunction;
    }
    FieldStringifier.prototype.isEmpty = function (value) {
        return typeof value === 'undefined' || value === null || value === '';
    };
    FieldStringifier.prototype.quoteField = function (field) {
        return "\"".concat(field.replace(/"/g, '""'), "\"");
    };
    FieldStringifier.prototype.filterValue = function (value) {
        return this.filterFunction(value);
    };
    return FieldStringifier;
}());
exports.FieldStringifier = FieldStringifier;
var DefaultFieldStringifier = /** @class */ (function (_super) {
    __extends(DefaultFieldStringifier, _super);
    function DefaultFieldStringifier() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultFieldStringifier.prototype.stringify = function (value) {
        var filteredValue = this.filterValue(value);
        if (this.isEmpty(filteredValue)) {
            if (this.quoteEmptyFields)
                return '""';
            return '';
        }
        var str = String(filteredValue);
        return this.needsQuote(str) ? this.quoteField(str) : str;
    };
    DefaultFieldStringifier.prototype.needsQuote = function (str) {
        return (str.includes(this.fieldDelimiter) ||
            str.includes('\r') ||
            str.includes('\n') ||
            str.includes('"'));
    };
    return DefaultFieldStringifier;
}(FieldStringifier));
var ForceQuoteFieldStringifier = /** @class */ (function (_super) {
    __extends(ForceQuoteFieldStringifier, _super);
    function ForceQuoteFieldStringifier() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ForceQuoteFieldStringifier.prototype.stringify = function (value) {
        var filteredValue = this.filterValue(value);
        if (this.isEmpty(filteredValue)) {
            if (this.quoteEmptyFields)
                return '""';
            return '';
        }
        var str = String(filteredValue);
        return this.quoteField(str);
    };
    return ForceQuoteFieldStringifier;
}(FieldStringifier));
function createFieldStringifier(fieldDelimiter, alwaysQuote, quoteEmptyFields, filterFunction) {
    if (fieldDelimiter === void 0) { fieldDelimiter = exports.DEFAULT_FIELD_DELIMITER; }
    if (alwaysQuote === void 0) { alwaysQuote = false; }
    if (quoteEmptyFields === void 0) { quoteEmptyFields = false; }
    if (filterFunction === void 0) { filterFunction = function (value) { return value; }; }
    _validateFieldDelimiter(fieldDelimiter);
    return alwaysQuote
        ? new ForceQuoteFieldStringifier(fieldDelimiter, quoteEmptyFields, filterFunction)
        : new DefaultFieldStringifier(fieldDelimiter, quoteEmptyFields, filterFunction);
}
function _validateFieldDelimiter(delimiter) {
    if (exports.VALID_FIELD_DELIMITERS.indexOf(delimiter) === -1) {
        throw new Error("Invalid field delimiter `".concat(delimiter, "` is specified"));
    }
}
